osu.Game/Screens/SelectV2/BeatmapCarouselFilterGrouping.cs

```diff
@@ -36,137 +35,115 @@ public BeatmapCarouselFilterGrouping(Func<FilterCriteria> getCriteria)
             this.getCriteria = getCriteria;
         }
 
-        public async Task<IEnumerable<CarouselItem>> Run(IEnumerable<CarouselItem> items, CancellationToken cancellationToken) => await Task.Run(() =>
+        public async Task<IEnumerable<CarouselItem>> Run(IEnumerable<CarouselItem> items, CancellationToken cancellationToken)
         {
-            setItems.Clear();
-            groupItems.Clear();
-
-            var criteria = getCriteria();
-            var newItems = new List<CarouselItem>(items.Count());
-
-            // Add criteria groups.
-            switch (criteria.Group)
+            return await Task.Run(() =>
             {
-                default:
-                    BeatmapSetsGroupedTogether = true;
-                    newItems.AddRange(items);
-                    break;
+                setItems.Clear();
+                groupItems.Clear();
 
-                case GroupMode.Artist:
-                    BeatmapSetsGroupedTogether = true;
-                    char groupChar = (char)0;
-
-                    foreach (var item in items)
-                    {
-                        cancellationToken.ThrowIfCancellationRequested();
+                var criteria = getCriteria();
+                var newItems = new List<CarouselItem>();
 
-                        var b = (BeatmapInfo)item.Model;
+                BeatmapInfo? lastBeatmap = null;
+                GroupDefinition? lastGroup = null;
 
-                        char beatmapFirstChar = char.ToUpperInvariant(b.Metadata.Artist[0]);
+                HashSet<CarouselItem>? groupRefItems = null;
+                HashSet<CarouselItem>? setRefItems = null;
```

Comment:
The `Ref` infix is majorly throwing me off here, can this be renamed to something better? If I understand correctly something like `current{Group,Set}Items` might work?

Reply:
Have changed these.